src/GeolocationContext.tsx

```diff
@@ -143,39 +152,46 @@ export const GeolocationContextProvider: React.FC = ({children}) => {
     }
   }
 
+  const watchIdRef = useRef<number>(-1);
+
   const hasPermission = state.status === 'granted' && state.locationEnabled;
-  useEffect(() => {
-    let watchId: number;
 
-    async function startLocationWatcher() {
-      if (!hasPermission) {
-        dispatch({
-          type: 'LOCATION_CHANGED',
-          position: null,
-          locationName: geoLocationName,
-        });
-      } else {
-        const config: GeoOptions = {
-          enableHighAccuracy: true,
-          distanceFilter: 20,
-        };
+  const startLocationWatcher = useCallback(async () => {
+    if (!hasPermission) {
+      dispatch({
+        type: 'LOCATION_CHANGED',
+        position: null,
+        locationName: geoLocationName,
+      });
+    } else {
+      const config: GeoOptions = {
+        enableHighAccuracy: true,
+        distanceFilter: 20,
+      };
 
-        watchId = Geolocation.watchPosition(
-          (position) => {
-            dispatch({
-              type: 'LOCATION_CHANGED',
-              position,
-              locationName: geoLocationName,
-            });
-          },
-          handleLocationError,
-          config,
-        );
-      }
+      watchIdRef.current = Geolocation.watchPosition(
+        (position) => {
+          dispatch({
+            type: 'LOCATION_CHANGED',
+            position,
+            locationName: geoLocationName,
+          });
+        },
+        handleLocationError,
+        config,
+      );
     }
+  }, [hasPermission, geoLocationName, dispatch]);
+
+  const stopLocationWatcher = () => Geolocation.clearWatch(watchIdRef.current);
+
+  useEffect(() => {
     startLocationWatcher();
-    return () => Geolocation.clearWatch(watchId);
-  }, [hasPermission]);
+    return stopLocationWatcher;
+  }, []);
```

Comment:
`startLocationWatcher` must be in dependency list so changes in `hasPermission` and `getLocationName` have effect?

Reply:
Yes, good catch!